src/GafferUIModule/PathListingWidgetBinding.cpp

```diff
@@ -1522,6 +1539,44 @@ class PathModel : public QAbstractItemModel
 			{
 				expansionChanged();
 			}
+
+			if( QGuiApplication::keyboardModifiers() & Qt::ShiftModifier )
+			{
+				// Recursively expand everything below this index. This
+				// could be expensive, so we do it during the async update.
+				cancelUpdate();
+				m_recursiveExpansionPath = expandedPath;
+				static_cast<Item *>( index.internalPointer() )->dirtyExpansion();
+				scheduleUpdate();
+			}
+		}
+
+		void finaliseRecursiveExpansion()
+		{
+			if( !m_recursiveExpansionPath )
+			{
+				return;
+			}
+			// We've expanded a bunch of paths on the Qt side, and now
+			// we need to reflect that in `m_expandedPaths`.
```

Comment:
> Do we not have a good way of getting a PathMatcher for all locations under m_recursiveExpansionPath without referencing QT though?

We can get it by calling `path->children()` recursively, but that could be slow, so we want to benefit from the fact that we've already done all that work in the background. We could add each path to `m_expandedPaths` in the same edit we use to expand it on the Qt side, but I had deliberately avoided that. I thought it was better, or at least more predictable initially, to avoid a huge flurry of `expansionChangedSignal()` that would cause repeated cancellations and restarts in the Viewer.

Reply:
That makes sense.